admin/aaxcvt.1.in
.\" Manpage for aaxcvt.
.\" Contact tech@adalin.com to correct errors or typos.
.TH man 1 "19 Dec 2014" "@AAX_UTILS_MAJOR_VERSION@.@AAX_UTILS_MINOR_VERSION@.@AAX_UTILS_MICRO_VERSION@" "aaxcvt man page"
.SH NAME
aaxcvt \- Converts a WAV audio file to a new format
.SH SYNOPSIS
.B aaxcvt
[\fIOPTION\fR]...
.SH DESCRIPTION
.PP
Converts a WAV input audio file to an output file in the the specified output format.
.TP
\fB\-i\fR, \fB\-\-input \fRFILE\fR
convert audio from this WAV file
.TP
\fB\-o\fR, \fB\-\-output \fRFILE\fR
write the audio to this file
.TP
\fB\-r\fR, \fB\-\-raw
do not write the WAV file header if specified
.TP
\fB\-f\fR, \fB\-\-format \fRFORMAT\fR
specifies the output format
.TP
\fB\-l\fR, \fB\-\-list
show a list of all supported formats
.TP
\fB\-h\fR, \fB\-\-help
print this message and exit
.PP
NOTE: WAV files are little endian only and AeonWave automatically compensates for that.
.SH AUTHOR
Written by Erik Hofman <tech@adalin.com>
.SH SEE ALSO
aaxinfo(1), aaxplay(1)



ChangeLog
* Fri Dec 12 2014 - tech@adalin.org
- AeonWave (2.5.0) stable; urgency=low
 * Added a capabilities interface to detect the limits of the audio hardware
   by making calls to the aaxMixerGetSetup function.
 * Changed the threading model from one thread per mixer-registered audio-frame
   to a load-balancing threadpool model. This slightly increases the work the
   CPU has to do but devides the wall time spent on rendering by the number of
   physical cores. Threading is now always active on multi-core systems except
   when the environment variable AAX_USE_THREADPOOL is set to 0.
 * Add support for the ARM architecture including support for hardware FPU
   and NEON. Rendering performance is comparible with Intel/AMD Cores
   at a matching CPU frequency.
 * Support batched (fast, non-realtime) mode for Audio Files and Loopback.
 * Support defining a default output device in the confguration file.
 * Fix a problem when a registered sensor had a different number of
   audio periods in relation to the playback device.
 * Fix an emitter looping bug introduced in version 2.4.1

 -- Erik Hofman <tech@adalin.org>  Fri, 12 Dec 2014 14:16:00 + 0100

* Tue Jun 10 2014 - tech@adalin.org
- AeonWave (2.4.1) stable; urgency=low
 * Fixed a possible segmentation fault with certain ALSA drivers.
 * Fixed a possible segmentation fault for looping emitters.

 -- Erik Hofman <tech@adalin.org>  Tue, 10 Jun 2014 12:58:00 + 0200

* Wed May 21 2014 - tech@adalin.org
- AeonWave (2.4.0) stable; urgency=low
 * Changed the conditions for binary distribution;
   The library may now be included in commercial and non commercial projects
   provided the user is made aware of the option to upgrade to the HD version.
   See Copyright section 2.1; Rights and Limitations of Grant.
 * Reorganized the code to support different hardware rendering schemes in
   the future. As a result the current rendering speed has increased 25% to 75%
   (depending of the computer hardware, older hardware profit the most).
 * Add the option to use timer-based rendering instead of interrupt-based
   rendering for ALSA (experimental).
 * Fix a problem where the WASAPI device period could become lower than the 
   minimum allowed period for that particular device.
 * Switched to a new xml configuration file format.
   Both the old and the new formats will be supported up to version 3.0.0
 * Allow custom pcm devices (defined in .asoundrc) for ALSA.
 * Tweaked Surround Sound, Spatial and HRTF playback.
 * Support WAV files with a LIST section.
 * Fix Auto generated waveform loop-points.
 * Fix a problem with capture-only backends

 -- Erik Hofman <tech@adalin.org>  Wed, 21 Mar 2014 10:45:00 +0200

* Sat Dec 21 2013 - tech@adalin.org
- AeonWave (2.3.4) stable; urgency=high
 * Fix a problem where distance delayed objects that where registered at
   the mixer or an audio-frame in non-playing mode started with the wrong 
   position and orientation matrix causing unwanted delays.
 * Fix an incorrect header problem while streaming mp3 files.
 * Fix a stupid error where AeonWave-Lite didn't initialize correctly.
 * Fix volume handling for windows devices without hardware volume support.
 * Activate AVX instruction support.

 -- Erik Hofman <tech@adalin.org>  Sat, 21 Dec 2013 11:21:00 +0100

* Mon Nov 25 2013 - tech@adalin.org
- AeonWave (2.3.3) stable; urgency=medium
 * Fix a problem with files with a .wav extension that aren't.
 * Prevent destroying of the emitter unless it is in the AAX_PROCESSED state.

 -- Erik Hofman <tech@adalin.org>  Mon, 25 Nov 2013 12:34:00 +0100

* Tue Nov 21 2013 - tech@adalin.org
- AeonWave (2.3.2) stable; urgency=medium
 * Fix clipping during spoken text under certain conditions.
 * Fix a problem where the equalizer was turned off under certain conditions.
 * Fix incorrect file-type handling for the Audio Files backend and add proper
   error reporting to the calling program.

 -- Erik Hofman <tech@adalin.org>  Tue, 21 Nov 2013 13:21:00 +0100

* Sun Nov 14 2013 - tech@adalin.org
- AeonWave (2.3.1) stable; urgency=high
 * Complete rewrite of the File backend
  - Add mp3 file output support if libmp3lame is found.
  - Add support for more WAV file types and fix mulaw and alaw decoding.
 * Add the option to register one sensor opened in write mode (file only) at
   the mixer to support a dedicated record to file option.
 * Fix a major registered-sensor bug (most noticeable on Windows).
 * Fix an occasional deadlock when deregistering audio-frames.

 -- Erik Hofman <tech@adalin.org>  Sun, 14 Nov 2013 15:14:00 +0200

* Fri Oct 18 2013 - tech@adalin.org
- AeonWave (2.3.0) stable; urgency=high
 * Distance-delay mode now adds delayed state changes in addition to delayed
   start/stop. State changes include; pitch, gain, position, orientation and
   velocity.
 * Added sense of height to Surround sound; emitters can now be heard above
   and below the speakers. This is best perceived if the speakers and the
   listeners ears are (almost) in the same horizontal plane.
 * Add support for the AAXS (AeonWave Auto-generated XML Sound) audio format.
 * Add double precision 64-bit floating point matrix support functions.
 * Implement hardware volume support for ALSA and WASAPI.
   - hardware volume is adjusted within hardware boundaries, volume settings
     outside the hardware range are adjusted using software fall-back.
   - the volume filter now reports the hardware minimum and maximum.
 * Implement Auto-Gain Control with a user definable adjust-rate.
 * Add the option to select stereo, mono-left, mono-right or mono-mix for
   audio capturing.
 * Add an aaxMixerGetState function to retrieve the state of the current
   audio stream. This can for instance be used to reinitialize the device
   after the hardware audio format has changed.
 * Add the option to set the number of times to loop a buffer.
 * Fixed reported bugs.
   - Fix an issue that prevented ALSA from reporting new hardware while an
     application was running.

src/aaxcvt.c
/*
 * Copyright (C) 2014 by Erik Hofman.
 * Copyright (C) 2014 by Adalin B.V.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *    1. Redistributions of source code must retain the above copyright notice,
 *        this list of conditions and the following disclaimer.
 * 
 *    2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY ADALIN B.V. ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
 * NO EVENT SHALL ADALIN B.V. OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUTOF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of Adalin B.V.
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>

#include <aax/aax.h>

#include "base/types.h"
#include "driver.h"
#include "wavfile.h"

#define MAX_LOOPS		6
static int _mask_t[MAX_LOOPS] = {
    0,
    AAX_FORMAT_UNSIGNED,
    AAX_FORMAT_LE,
    AAX_FORMAT_LE_UNSIGNED,
    AAX_FORMAT_BE,
    AAX_FORMAT_BE_UNSIGNED
};

static const char* _format_s[][2] = {
    { " AAX_PCM8S", "\tsigned, 8-bits per sample" },
    { " AAX_PCM16S", "\tsigned, 16-bits per sample" },
    { " AAX_PCM24S", "\tsigned, 24-bits per sample, 32-bit encoded" },
    { " AAX_PCM32S", "\tsigned, 32-bits per sample" },
    { " AAX_FLOAT", "\t32-bit floating point, -1.0f to 1.0f" },
    { " AAX_DOUBLE", "\t64-bit floating point, -1.0f to 1.0f" },
    { " AAX_MULAW", "\tmulaw, 16-bit with 2:1 compression" },
    { " AAX_ALAW", "\talaw, 16-bit with 2:1 compression" },
    { " AAX_IMA4_ADPCM", "IMA4 ADPCM, 16-bit with 4:1 compression" }
};

static const char* _format_us[][2] = {
    { " AAX_PCM8U", "\tunsigned, 8-bits per sample" },
    { " AAX_PCM16U", "\tunsigned, 16-bits per sample" },
    { " AAX_PCM24U", "\tunsigned, 24-bits per sample, 32-bit encoded" },
    { " AAX_PCM32U", "\tunsigned, 32-bits per sample" }
};

static const char* _mask_s[MAX_LOOPS][2] = {
    { "Native format", ":\t" },
    { "Unigned format", ":\t" },
    { "Little endian, signed", "_LE:" },
    { "Little endian, unsigned", "_LE:" },
    { "Big endian, signed", "_BE:" },
    { "Big endian, unsigned", "_BE:" }
};

void
help()
{
    printf("aaxcvt version %i.%i.%i\n\n", AAX_UTILS_MAJOR_VERSION,
                                           AAX_UTILS_MINOR_VERSION,
                                           AAX_UTILS_MICRO_VERSION);
    printf("Usage: aaxcvt [options]\n");
    printf("Converts an input WAV audio file to an output file in the the"
           "\nspecified output format.\n");

    printf("\nOptions:\n");
    printf("  -i, --input <file>\t\tconvert audio from this WAV file\n");
    printf("  -o, --output <file>\t\twrite the audio to this file\n");
    printf("  -r, --raw\t\t\tdo not write the WAV file header if specified\n");
    printf("  -f, --format <format>\t\tspecifies the output format\n");
    printf("  -l, --list\t\t\tshow a list of all supported formats\n");
    printf("  -h, --help\t\t\tprint this message and exit\n");

    printf("\nNote that WAV files are little endian only and AeonWave "
           "automatically\ncompensates for that.\n");

    printf("\n");
    exit(-1);
}

void
list()
{
    int q;
    for (q=0; q<MAX_LOOPS; q++)
    {
        int fmt = 0;
        printf("%s\n", _mask_s[q][0]);
        do
        {
            int nfmt = fmt + _mask_t[q];
 
            if (q)
            {
                if (q > 1 && fmt == 0) {
                    continue;
                }

                if (fmt >= AAX_FLOAT && nfmt & AAX_FORMAT_UNSIGNED) {
                    break;
                }
                if (fmt >= AAX_MULAW) {
                    break;
                }
            }

            if (nfmt & AAX_FORMAT_UNSIGNED)
            {
                printf("  %s%s%s\n", _format_us[fmt & AAX_FORMAT_NATIVE][0],
                                       _mask_s[q][1],
                                       _format_us[fmt & AAX_FORMAT_NATIVE][1]);
            }
            else
            {
                printf("  %s%s%s\n", _format_s[fmt & AAX_FORMAT_NATIVE][0],
                                       _mask_s[q][1],
                                       _format_s[fmt & AAX_FORMAT_NATIVE][1]);
            }
        }
        while (++fmt < AAX_FORMAT_MAX);
    }
    exit(-1);
}

int main(int argc, char **argv)
{
    enum aaxFormat format;
    char *infile, *outfile;
    int raw, rv = 0;

    if (getCommandLineOption(argc, argv, "-l") ||
        getCommandLineOption(argc, argv, "--list"))
    {
       list();
    }

    if ((argc < 7) || (argc > 8))  {
        help();
    }

    raw = getCommandLineOption(argc, argv, "-r") ? 1 : 0;

    infile = getInputFile(argc, argv, NULL);
    if (!infile)
    {
        printf("Input file not specified or unable to access.\n");
        help();
        exit(-2);
    }
    outfile = getOutputFile(argc, argv, NULL);
    if (!outfile) {
       help();
    }

    format = getAudioFormat(argc, argv, AAX_FORMAT_NONE);
    if (format != AAX_FORMAT_NONE)
    {
        aaxConfig config;
        config= aaxDriverOpenByName("AeonWave Loopback", AAX_MODE_WRITE_STEREO);
        aaxBuffer buffer = bufferFromFile(config, infile);
        if (buffer)
        {
            aaxBufferSetSetup(buffer, AAX_FORMAT, format);
            if (!raw) {
               aaxBufferWriteToFile(buffer, outfile, AAX_OVERWRITE);
            }
            else
            {
               int fd = open(outfile, O_CREAT|O_WRONLY, 0644);
               if (fd >= 0)
               {
                  int size = aaxBufferGetSetup(buffer, AAX_TRACKSIZE);
                  void **data = aaxBufferGetData(buffer);
                  int res = write(fd, *data, size);
                  if (res != size) {
                     printf("Written %i bytes of the required: %i\n", res, size);
                  }
                  close(fd);
               }
               else {
                  printf("Unable to open file for writing: %s\n", outfile);
               }
            }
            aaxBufferDestroy(buffer);
        }
        aaxDriverDestroy(config);
    }
    else {
        printf("Unsupported audio format.\n");
        rv = -2;
    }

    return rv;
}

